src/manifest.rs

```diff
@@ -0,0 +1,442 @@
+use proto::meta::{
+    manifest_change::Operation as ManifestChangeOp, ManifestChange, ManifestChangeSet,
+};
+
+use crate::{util, write_with_length_check, AgateOptions, Error, Result};
+
+pub const MANIFEST_FILENAME: &str = "MANIFEST";
+const MANIFEST_REWRITE_FILENAME: &str = "MANIFEST_REWRITE";
+const MANIFEST_DELETION_REWRITE_THRESHOLD: usize = 10000;
+const MANIFEST_DELETIONS_RATIO: usize = 10;
+
+const MAGIC_TEXT: &[u8] = b"Agat";
+const MAGIC_VERSION: u32 = 8;
+
+/// LevelManifest contains information about LSM tree levels
+/// in the MANIFEST file.
+#[derive(Default, Clone, Debug)]
+pub struct LevelManifest {
+    pub tables: HashSet<u64>,
+}
+
+/// TableManifest contains information about a specific table
+/// in the LSM tree.
+#[derive(Clone, Debug)]
+pub struct TableManifest {
+    pub level: u8,
+    pub key_id: u64,
+    // TODO: compression
+}
+
+/// Manifest represents the contents of the MANIFEST file.
+#[derive(Clone, Debug)]
+pub struct Manifest {
+    pub levels: Vec<LevelManifest>,
+    pub tables: HashMap<u64, TableManifest>,
+    pub creations: usize,
+    pub deletions: usize,
+}
+
+struct Core {
+    file: Option<File>,
+    manifest: Manifest,
+}
+
+impl Core {
+    fn rewrite(&mut self, dir: &Path) -> Result<()> {
+        self.file.take();
+        let (file, net_creations) = ManifestFile::help_rewrite(dir, &self.manifest)?;
+        self.file = Some(file);
+        self.manifest.creations = net_creations;
+        self.manifest.deletions = 0;
+
+        Ok(())
+    }
+}
+
+// ManifestFile holds the file pointer (and other info) about the MANIFEST file, which is a log
+// file we append to.
+pub struct ManifestFile {
+    directory: PathBuf,
+    deletions_rewrite_threshold: usize,
+    core: Mutex<Core>,
+}
+
+impl Manifest {
+    pub fn new() -> Self {
+        Self {
+            levels: vec![],
+            tables: HashMap::new(),
+            creations: 0,
+            deletions: 0,
+        }
+    }
+
+    fn as_changes(&self) -> Vec<ManifestChange> {
+        let mut changes = Vec::with_capacity(self.tables.len());
+        for (id, tm) in &self.tables {
+            changes.push(new_create_change(*id, tm.level as usize, tm.key_id));
+        }
+        changes
+    }
+
+    pub fn replay(file: &mut File) -> Result<(Manifest, u32)> {
+        file.seek(SeekFrom::Start(0))?;
+        let mut magic_buf = vec![0; 8];
+        file.read_exact(&mut magic_buf)?;
+        if &magic_buf[..4] != MAGIC_TEXT {
+            return Err(Error::CustomError("bad magic text".to_string()));
+        }
+        let version = (&magic_buf[4..]).get_u32();
+        if version != MAGIC_VERSION {
+            return Err(Error::CustomError("bad magic version".to_string()));
+        }
+
+        let stat = file.metadata()?;
+        let mut build = Manifest::new();
+        let mut buf = vec![];
+        let mut len_crc_buf = vec![0; 8];
+        let mut offset = 8;
+
+        loop {
+            if file.read(&mut len_crc_buf)? != 8 {
```

Comment:
I guess read function is not guaranteed to read 8 bytes. Maybe we should use `read_exact` here, and find EOF error to break.